src/System.Private.Uri/src/System/Uri.cs

```diff
@@ -3928,8 +3928,60 @@ unsafe static private ParsingError CheckSchemeSyntax(char* ptr, ushort length, r
                     return ParsingError.BadScheme;
                 }
             }
-            // A not well-known scheme, needs string creation
-            // Note it is already in the lower case as required.
+
+            // Special-case common and known schemes to avoid allocations and dictionary lookups in these cases.
+            switch (length)
+            {
+                case 2:
+                    if (ptr[0] == 'w' && ptr[1] == 's')
+                    {
+                        syntax = UriParser.WsUri;
+                        return ParsingError.None;
+                    }
+                    break;
+                case 3:
+                    const int ftpMask = 'f' << 16 | ('t' << 8) | 'p';
```

Comment:
I went back and forth on that.  On the one hand, it's the guideline.  On the other, this maps better to the exact string/char set that's being used, e.g. "ftp" instead of "Ftp".  And since the characters here exactly what's being compared and are important, I thought the clarity provided by that outweighed the consistency of following the guideline.